be/src/storage/predicate_tree/predicate_tree.h

```diff
@@ -0,0 +1,357 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <memory>
+#include <set>
+#include <vector>
+
+#include "common/overloaded.h"
+#include "storage/column_predicate.h"
+#include "storage/predicate_tree/predicate_tree_fwd.h"
+
+namespace starrocks {
+
+class ColumnPredicate;
+
+// ------------------------------------------------------------------------------------
+// Utils
+// ------------------------------------------------------------------------------------
+
+using PredicateColumnNodes = std::vector<PredicateColumnNode>;
+using PredicateColumnNodeMap = std::unordered_map<ColumnId, PredicateColumnNodes>;
+
+using PredicateNodeId = uint32_t;
+using ColumnPredicates = std::vector<const ColumnPredicate*>;
+using ColumnPredicateMap = std::unordered_map<ColumnId, ColumnPredicates>;
+
+template <typename T>
+concept PredicateNodeType = std::is_same_v<T, PredicateColumnNode> || std::is_same_v<T, PredicateAndNode> ||
+                            std::is_same_v<T, PredicateOrNode>;
+
+// ------------------------------------------------------------------------------------
+// PredicateTreeNode Nodes
+// ------------------------------------------------------------------------------------
+
+class PredicateBaseNode {
+public:
+    PredicateNodeId id() const { return _id; }
+    void set_id(PredicateNodeId id) { _id = id; }
+
+protected:
+    /// The id of any compound node is placed before the id of any column node.
+    /// The id of each node is set after the PredicateTree is constructed.
+    PredicateNodeId _id = 0;
+};
+
+template <typename Derived>
+class PredicateBaseNodeHelper : public PredicateBaseNode {
```

Comment:
resembles to FixedLengthColumnBase : public ColumnFactory, can we use PredicateBaseNodeFactory instead of PredicateBaseNodeHelper?